tss-esapi/src/abstraction/web_authn.rs

```diff
@@ -0,0 +1,192 @@
+            "pubArea" => pub_area,
+            "certInfo" => cert_info,
+        }) {
+            Ok(value) => match into_writer(&value, &mut encoded_token) {
+                Ok(_) => Ok(encoded_token),
+                Err(_) => Err(Error::local_error(InternalError)),
+            },
+            Err(_) => Err(Error::local_error(InternalError)),
+        }
+    }
+
+    fn get_alg_value(algorithm: SignatureScheme) -> i32 {
+        match algorithm {
+            SignatureScheme::RsaSsa { .. } => -65535,
+            _ => -1,
+        }
+    }
+}
+
+#[derive(Debug)]
+pub struct TpmPlatStmt {
+    attestation: Attest,
+    signature: Signature,
+    algorithm: SignatureScheme,
+    x509_chain: Vec<Vec<u8>>,
+}
+
+impl TpmPlatStmt {
+    pub fn new(
+        context: &mut Context,
+        key: KeyHandle,
+        nonce: Vec<u8>,
+        selection_list: PcrSelectionList,
+    ) -> Result<TpmPlatStmt> {
+        let session_1 = context.start_auth_session(
+            None,
+            None,
+            None,
+            SessionType::Hmac,
+            SymmetricDefinition::AES_128_CFB,
+            HashingAlgorithm::Sha256,
+        )?;
+        let (key_public_area, _, _) = context.read_public(key)?;
+        let signing_scheme = get_sig_scheme(key_public_area)?;
+        let (attestation, signature) = context
+            .execute_with_sessions((session_1, None, None), |ctx| {
+                ctx.quote(key, nonce.try_into()?, signing_scheme, selection_list)
+            })
+            .or_else(|e| {
+                context.flush_context(SessionHandle::from(session_1).into())?;
+                Err(e)
+            })?;
+        context.flush_context(SessionHandle::from(session_1).into())?;
+        Ok(TpmPlatStmt {
+            attestation,
+            signature,
+            x509_chain: Vec::new(),
+            algorithm: signing_scheme,
+        })
+    }
+
+    pub fn add_certificates(&mut self, mut certificates: Vec<Vec<u8>>) {
+        self.x509_chain.append(&mut certificates);
+    }
+
+    pub fn encode(&self) -> Result<Vec<u8>> {
+        let sig = self.signature.clone().marshall()?;
+        let cert_info = self.attestation.clone().marshall()?;
+        let alg = TpmPlatStmt::get_alg_value(self.algorithm);
+        let mut encoded_token = vec![];
+        match cbor!({
+            //TODO put brakets around x5c and alg
+            "x5c" => self.x509_chain.clone(),
+            "alg" => alg,
+            "sig" => sig,
+            "certInfo" => cert_info,
+        }) {
+            Ok(value) => match into_writer(&value, &mut encoded_token) {
+                Ok(_) => Ok(encoded_token),
+                Err(_) => Err(Error::local_error(InternalError)),
+            },
+            Err(_) => Err(Error::local_error(InternalError)),
+        }
+    }
+
+    fn get_alg_value(algorithm: SignatureScheme) -> i32 {
+        match algorithm {
+            SignatureScheme::RsaSsa { .. } => -65535,
+            _ => -1,
+        }
+    }
```

Comment:
Can probably do with this the same as with `get_sig_scheme` and just take it out as an independent function.